Repository/CompositeRepository.php

```diff
@@ -0,0 +1,180 @@
+<?php
+
+namespace Symfony\Cmf\Component\Resource\Repository;
+
+use InvalidArgumentException;
+use Puli\Repository\Api\ResourceCollection;
+use Puli\Repository\Api\ResourceNotFoundException;
+use Puli\Repository\Api\ResourceRepository;
+use Puli\Repository\Resource\Collection\ArrayResourceCollection;
+use Webmozart\PathUtil\Path;
+use Webmozart\Assert\Assert;
+use Puli\Repository\InMemoryRepository;
+
+/**
+ * Composite repository.
+ *
+ * Aggregates multiple repositories into one.
+ *
+ * - Repositories can be mounted at "/" or as immediate children of "/".
+ * - Repositories can be mounted within other repositories.
+ *
+ * ````
+ * $compositeRepository->mount('/', $repository1);
+ * $compositeRepository->mount('/foo', $repository2);
+ * $resources = $compositeRepository->listChildren('/');
+ * var_dump($resource->getPaths()); // will list all children of repository 1 and also a /foo node
+ * ````
+ *
+ * In the case that a mounted repository conflicts with a resource in an existing repository,
+ * the mounted repository will take precedence.
+ */
+class CompositeRepository implements ResourceRepository
+{
+    private $repos = array();
+
+    /**
+     * Mount a repository at the given path.
+     *
+     * The path should be either "/" or an immediate child of "/" or
+     * an immediate child of a resource in an existing mounted repository.
+     *
+     * @param string $path
+     * @param ResourceRepository $repository
+     */
+    public function mount($path, ResourceRepository $repository)
+    {
+        $this->repos[$path] = $repository;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function get($path)
+    {
+        list($repository, $repoPath, $path) = $this->getRepository($path);
```

Comment:
It occured to me that we could create a wrapped repository which is aware that it is mounted and contains a reference to the mount point. This would illiminate the need for returning a tuple. But, probably not worth the effort.